### PR TITLE
Create database-pool image

### LINKED ISSUE #308: Database Pool: Create docker image and update ubuntu-base
We want to move away from including every database in the base image and instead create a single image including every database server.

This task consists of two parts:

1. [create database-pool image](https://codesignal.atlassian.net/wiki/spaces/PE/pages/797507585/Run+database+containers+separately+from+language-environment+containers#Docker-image-specification-for-database-container)
1.1 install every supported database server
1.2 create shell scripts allowing to start the corresponding database server
1.3 create simple HTTP server allowing to execute corresponding shell script)
2. [update ubuntu-base image](https://codesignal.atlassian.net/wiki/spaces/PE/pages/797507585/Run+database+containers+separately+from+language-environment+containers#Docker-image-specification-for-the-base-image)
2.1 install CLI tools for each supported database server
2.2 create shell scripts that let the users start a database server by sending an HTTP request to the database container

### PR DESCRIPTION
**Pull request by mpowaga**
Closes #308 
### Overview

This PR adds a new docker image (codesignal/database-pool) and updates ubuntu-base image.

~CI fails because we need to create a codesignal/database-pool repository in dockerhub.~
~codesignal/database-pool repository is created in dockerhub but CI still fails to build because build time is exceeded (image size is \~3GB so the build and upload take longer than the time limit).~

This PR should be merged together with [coderunner/#1061](https://github.com/CodeSignal/coderunner/issues/1061) (probably there will be some small changes to this PR affected by coderunner changes).

### Some ~open~ questions:

1. What should be the default port for the database-pool HTTP server? It is configurable but since this server is used only internally we should set it to something that won't likely be used by the users.

I decided to remove the default port since we always pass it as an env variable. In the coderunner we set it to `20215`.

2. Where database run scripts should be placed? Right now it's `/etc/database-pool` but maybe there is a better directory? Do note that these scripts will be used by users in `main.sh` (maybe we can put them in `/usr/local/bin` or some other bin directory).

Seems like this directory is ok since nobody objected.

3. Do we need to allow root access for the users? Since now connections will be remote (not through localhost) the root access is disabled for most database servers (only Microsoft SQL Server seems to allow it with no change in configuration).

Most likely not but will check with Dima.

### Testing

The script that I used for testing.

```sh
#!/usr/bin/env sh

TAG="dev"
NETWORK="db"
DB_POOL_HOST="database-pool"
DB_POOL_PORT="8080"

docker build -t codesignal/ubuntu-base:$TAG ./base
docker build -t codesignal/database-pool:$TAG ./database-pool

docker network inspect $NETWORK >/dev/null 2>&1 || \
    docker network create --driver bridge $NETWORK

docker container inspect database-pool >/dev/null 2>&1 && \
    docker rm --force database-pool
docker run --network $NETWORK --name database-pool -e DB_POOL_PORT=$DB_POOL_PORT -d codesignal/database-pool:$TAG

cat <<EOF | docker run --rm -i --network=$NETWORK -e DB_POOL_HOST=$DB_POOL_HOST -e DB_POOL_PORT=$DB_POOL_PORT codesignal/ubuntu-base:$TAG sh
echo "\n\nTESTING MYSQL"
/etc/database-pool/mysql_start.sh
mysql ri_db -u'test' -h $DB_POOL_HOST -e "SELECT 102030;"

echo "\n\nTESTING POSTGRESQL"
/etc/database-pool/postgresql_start.sh
psql -U postgres -h $DB_POOL_HOST -d ri_db -c "SELECT 302010;"

echo "\n\nTESTING MICROSOFT SQL"
/etc/database-pool/mssql_start.sh
/opt/mssql-tools/bin/sqlcmd -S $DB_POOL_HOST -U SA -P StrongPass1 -q "SELECT 123456;"

echo "\n\nTESTING MONGODB"
/etc/database-pool/mongo_start.sh
mkdir -p /home/mongo_test
cd /home/mongo_test
npm install mongodb
cat - << END > index.js
const MongoClient = require('mongodb').MongoClient;
const assert = require('assert');

// Connection URL
const url = 'mongodb://' + process.env.DB_POOL_HOST + ':27017';

// Database Name
const dbName = 'myproject';
const client = new MongoClient(url);
// Use connect method to connect to the server
client.connect(function(err) {
  assert.equal(null, err);
  console.log('Connected successfully to server');

  const db = client.db(dbName);

  client.close();
});

END
node index.js

EOF

docker rm --force database-pool

```

### Migration plan

In order to enable the language environment to connect to the database, we need to update corresponding dockerfiles. It doesn't need to be done all at once. We can start with a limited number of languages first and then incrementally add more. 

Below is the instruction describing how to update a language-environment image to support database-pool.

#### STEP 1: Dockerfiles repo

1. Branch out from [database-pool-migration](https://github.com/CodeSignal/dockerfiles/tree/database-pool-migration)
2. Update language dockerfile to use the latest version of ubuntu-base image (with database-pool support) and install connector libraries for each database server (MySQL, PostgreSQL, MongoDB, Microsoft SQL Server)
3. Create a PR and wait for CI build to finish. It will push the image to docker hub with tag `new`

#### STEP 2: Coderunner repo

1. Update [compilers.ts](https://github.com/CodeSignal/coderunner/blob/master/API/src/compilers.ts) and [UpdateDocker.sh](https://github.com/CodeSignal/coderunner/blob/master/Setup/UpdateDocker.sh) to use an image with tag `new`
2. Create a PR and deploy a new coderunner from that branch
3. Make sure that all tests pass and coderunner is deployed successfully
4. Create `db.[language].test.js` tests that check the basic connection to all database servers (see [db.js.test.js](https://github.com/CodeSignal/coderunner/blob/2da7e4e0283f04515167e9b6d5dd095ab1e12ab2/API/test/db.js.test.js) for example)

#### STEP 3: Dockerfiles repo

1. Revert 65bf2c6 commit
2. Merge and deploy a new version of the image

#### STEP 4: Coderunner repo

1. Update [compilers.ts](https://github.com/CodeSignal/coderunner/blob/master/API/src/compilers.ts) and [UpdateDocker.sh](https://github.com/CodeSignal/coderunner/blob/master/Setup/UpdateDocker.sh) to use an image with the latest version
2. Redeploy a coderunner and check that tests are successful
3. Merge PR

### CODE REVIEW

**Comment by Albinutte** [comment]
@mpowaga I've created a repo on dockerhub, so you should now be able to push repositories to remote from the CI.

For the open questions you've listed:

1. Perhaps something not form the [list of commonly used ports](https://en.wikipedia.org/wiki/List_of_TCP_and_UDP_port_numbers)? In particular, I can imagine `8080` being used a front-end server in some tasks.
2. `/etc/database-pool` seems legit to me, I don't have a better suggestion there 😅 
3. Do you envision a scenario in which root access is necessary? Unless required, I'd prefer to stick to lower access level, but I don't think I fully understand all the use cases.

For the testing part, could you clarify how the script you provided can be used? Should it be used inside some container? What images should I have downloaded to run it?

Btw I love how much smaller the base image became:

image

**Review by salbert96 — COMMENTED** [review]
Leaving some comments from the first phase of the code review, but I did not test/experiment with it yet.

Can you also write what's our deployment strategy, are we going to have a separate branch and merge there all the changes?

**Comment by salbert96** [comment]
@mpowaga probably not relevant or solvable but we normally disable the network when running the containers
https://github.com/CodeSignal/coderunner/blob/master/API/DockerTimeout.sh

**Comment by salbert96** [comment]
I tested manually using your script. Pretty solid work @mpowaga great stuff 💪

**Comment by mpowaga** [comment]
@Albinutte CI fails with the following error when pushing `codesignal/database-pool` image:

> denied: requested access to the resource is denied

Do we need to configure permissions or something?

**Comment by Albinutte** [comment]
> Do we need to configure permissions or something?

Ah, you're right, sorry, forgot about that. Let me see if I can set it up.

Btw @menewman it looks like we've run out of free seats on dockerhub. Adding another user is just $9/month, so can we get one more?

**Comment by menewman** [comment]
Seat added and invite sent to `michal.powaga@codesignal.com`

File: .circleci/config.yml [hunk 1]
```diff
@@ -35,11 +35,12 @@ jobs:
                 fi
 
                 echo dockerfile: $IMAGE '->' repo: $REPO
-                docker build -t codesignal/$REPO:dev $IMAGE
-                docker push codesignal/$REPO:dev
+                docker build -t codesignal/$REPO:new $IMAGE
+                docker push codesignal/$REPO:new
```
Comment (salbert96) [inline .circleci/config.yml:39]: Tagging to make sure we don't merge it this way.

File: database-pool/scripts/mssql_start.sh [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+CREATE LOGIN test WITH PASSWORD = 'test', CHECK_POLICY = OFF;
+CREATE USER test FOR LOGIN test;
+GRANT CONTROL ON DATABASE::test TO test;
+-- sysadmin role is required for raiserror with log command
+-- which we use in the template to terminate the execution
+exec sp_addsrvrolemember 'test', 'sysadmin';
+ALTER LOGIN test ENABLE;
+GO
+END
+fi
```
Comment (salbert96) [inline database-pool/scripts/mssql_start.sh:38]: [minor] Can we add a new line here.

**Review by salbert96 — COMMENTED** [review]
@mpowaga what is the state of this PR, do you have blockers or action items on it? Because it seems like it is mostly ready. The only question I have is the MySQL version (since we had 2 versions before) maybe worth talking with Dima to make sure we don't break the existing MySQL tasks.

**Comment by mpowaga** [comment]
@salbert96 This PR is ready and there are no blockers. Will check with Dima about the MySQL version and root access.

**Comment by salbert96** [comment]
Seems like we also need to somehow change the base of database-pool from `new` 

I think ideally we would merge the base, create a release from it, then merge the database-pool, create a release from it, then all the other images based on ubuntu-base.

But I think we can't merge them separately so maybe it's okay to merge a nonworking version and then update it the base tag of the database-pool?

File: database-pool/pg_hba.conf [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+# IPv4 local connections:
+host    all             all             127.0.0.1/32            trust
+# IPv6 local connections:
+host    all             all             ::1/128                 trust
+# Allow replication connections from localhost, by a user with the
+# replication privilege.
+local   replication     all                                     peer
+host    replication     all             127.0.0.1/32            md5
+host    replication     all             ::1/128                 md5
+host    all             all             0.0.0.0/0               trust
```
Comment (salbert96) [inline database-pool/pg_hba.conf:104]: [minor] Can we have an end line here? 🙂

File: base/scripts/mongo_start.sh [hunk 4]
```diff
@@ -0,0 +1,8 @@
+#!/usr/bin/env sh
```
Comment (salbert96) [inline base/scripts/mongo_start.sh:1]: @mpowaga can we run `chmod 777 mongo_start.sh ` for these scripts?

File: .circleci/config.yml (hunk 1 above)
Comment (salbert96) [reply .circleci/config.yml:39]: @mpowaga let's remove these so that we can merge after approve.

File: database-pool/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,80 @@
+FROM codesignal/ubuntu-base:new
```
Comment (salbert96) [inline database-pool/Dockerfile:1]: After merging, we are going to need to push one more PR to change this to the new ubuntu-base.

**Comment by salbert96** [comment]
@mpowaga is the plan of merging the following.
1. Merge this PR
2. Create a new release for the ubuntu-base
3. Create a new PR and update the base of the database_pool
4. Merge it and create a release for the database pool
5. Update all the images as written in https://github.com/CodeSignal/dockerfiles#readme

**Comment by mpowaga** [comment]
@salbert96 I think that's a good plan. We could also create a separate PR with ubuntu-base changes which will be merged before this PR (so we can update the tag before merging).

**Comment by salbert96** [comment]
@mpowaga I think it's okay to have in the code with a `new` tag since it's temporary. It's not worth the struggle.

**Comment by salbert96** [comment]
@mpowaga please address the only comment after which I can approve and we can merge this.

**Review by salbert96 — APPROVED** [review]
Thanks for all the changes 🚀

**Pull request merged**